neuralogic/core/constructs/function/__init__.py

```diff
@@ -29,4 +32,78 @@
 Aggregation.SOFTMAX = Softmax("SOFTMAX")
 
 
-__all__ = ["Transformation", "Combination", "Aggregation", "Function"]
+class CombinationWrap:
+    __slots__ = "left", "right", "combination"
+
+    def __init__(self, left, right, combination: Combination):
+        self.left = left
+        self.right = right
+        self.combination = combination
+
+    def __add__(self, other):
+        return CombinationWrap(self, other, Combination.SUM)
+
+    def __mul__(self, other):
+        return CombinationWrap(self, other, Combination.ELPRODUCT)
+
+    def __matmul__(self, other):
+        return CombinationWrap(self, other, Combination.PRODUCT)
+
+    def __str__(self):
+        if not isinstance(self.left, CombinationWrap) and not isinstance(self.right, CombinationWrap):
+            return f"{self.combination}"
+        if not isinstance(self.left, CombinationWrap):
+            return f"{self.combination}({self.right.to_str()})"
+        if not isinstance(self.right, CombinationWrap):
+            return f"{self.combination}({self.left.to_str()})"
+
+        return f"{self.combination}({self.left.to_str()}, {self.right.to_str()})"
+
+    def __iter__(self):
+        if isinstance(self.left, CombinationWrap):
+            for a in self.left:
+                yield a
+        if not isinstance(self.left, CombinationWrap):
+            yield self.left
+
+        if isinstance(self.right, CombinationWrap):
+            for a in self.right:
+                yield a
+        if not isinstance(self.right, CombinationWrap):
+            yield self.right
+
+    def to_combination(self) -> Combination:
+        combination_graph, _ = self._get_combination_node()
+        return MixedCombination(name=self.to_str(), combination_graph=combination_graph)
+
+    def _get_combination_node(self, input_counter: int = 0):
+        left_node = None
+        right_node = None
+
+        left_index = -1
+        right_index = -1
+
+        if isinstance(self.left, CombinationWrap):
+            left_node, input_counter = self.left._get_combination_node(input_counter)
+        else:
+            left_index = input_counter
+            input_counter += 1
+
+        if isinstance(self.right, CombinationWrap):
+            right_node, input_counter = self.right._get_combination_node(input_counter)
+        else:
+            right_index = input_counter
+            input_counter += 1
+
+        class_name = "cz.cvut.fel.ida.algebra.functions.combination.MixedCombination.MixedCombinationNode"
+
+        return (
```

Comment:
Almost feel like there should be some simplification to combine all combinations into one if they are the same and the order is not changes.

That is, instead of sum(a, sum(b, sum(c, d))) have sum(a, b, c, d)

Reply:
Some combinations might not be permutation invariant - don't really want to deal with that, so do the simplification only when literals are in the original sequence?